Add "Combing through ComponentBase" blog post (#1)
captainsafia/blog: _posts/2020/2020-10-14-combing-through-component-base.md

Most? Not all? Correct! Not all components have to implement `ComponentBase`. In fact, the `Router` component that we looked at in last week's blog post doesn't derive from `ComponentBase`. It implements the much narrower `IComponent` interface ([ref](https://github.com/dotnet/aspnetcore/blob/795584e177852a2cb95d4e7b0df2ae7d94fac157/src/Components/Components/src/IComponent.cs)).

```csharp
public interface IComponent {
    void Attach(RenderHandle renderHandle);
    Task SetParametersAsync(ParameterView parameters);
}
```

Pretty bare bones, huh? Notably, the `IComponent` interface does not require that the component implement Blazor's component lifecycle methods or support any event handling. This makes the `IComponent` interface an ideal base for component's that don't require any kind of interactivity, like the `Router` component mentioned above or [the LayoutView component](https://github.com/dotnet/aspnetcore/blob/724c2e75a7415e67f9f5ad33fc72bf55507b8d59/src/Components/Components/src/LayoutView.cs).

> One key thing to note: the `IComponent` interface is used throughout the codebase for component discovery and rendering. That means tht component's aren't required to derive from `ComponentBase` in order to plug in with everything else.

Reviewer: ```suggestion
One key thing to note: the `IComponent` interface is used throughout the codebase for component discovery and rendering. That means that components aren't required to derive from `ComponentBase` in order to plug in with everything else.
```

```csharp
public abstract class ComponentBase : IComponent, IHandleEvent, IHandleAfterRender
```

This is only one line of code but we can actually unpack quite a bit from it.

For starters, `ComponentBase` is an `abstract` (aka base) class. You cannot instantiate an instance of the `ComponentBase` class itself, but most Blazor components derive from the `ComponentBase` class itself.

Most? Not all? Correct! Not all components have to implement `ComponentBase`. In fact, the `Router` component that we looked at in last week's blog post doesn't derive from `ComponentBase`. It implements the much narrower `IComponent` interface ([ref](https://github.com/dotnet/aspnetcore/blob/795584e177852a2cb95d4e7b0df2ae7d94fac157/src/Components/Components/src/IComponent.cs)).

```csharp
public interface IComponent {
    void Attach(RenderHandle renderHandle);
    Task SetParametersAsync(ParameterView parameters);
}
```

Pretty bare bones, huh? Notably, the `IComponent` interface does not require that the component implement Blazor's component lifecycle methods. We'll cover these lifecycle methods later in the chapter, but suffice to say that 

### The Lifecycle Methods

> Whenever I think about lifecycle methods in components (whether Blazor or React), I always think back to those "Life Cycle of a Cell" diagrams from high school. The concepts are actually fairly similar. A component has particular stages of development, from when it is first rendered on the page to when it is discarded. Components can also react to changes in their "environment", whether it is events being triggered or parameters changing.

Reviewer: ```suggestion
Whenever I think about lifecycle methods in components (whether Blazor or React), I always think back to those "Life Cycle of a Cell" diagrams from high school. The concepts are actually fairly similar. A component has particular stages of development, from when it is first rendered on the page to when it is discarded. Components can also react to changes in their "environment", such as triggered events or parameter changes.
```

#### SetParametersAsync

> After a component is initialized, we need to establish what parameters have been provided to the component. This needs to happen before any rendering occurs because what is rendered is often dependent on the parameters provided to the component.

Reviewer: Is there a more succinct to say "what parameters have been provided to the component"
